File: src/odd/ga_lib/include/coupledstrip_lib.cpp
// Include the header for this source file
#include "coupledstrip_lib.h"
#include <iostream>
#include <cmath>
#include <numbers>
#include <stdexcept>
#include <format>

/*
*******************************************************
*                      Constants                      *
*******************************************************
*/
constexpr double PI = std::numbers::pi;
constexpr double E0 = 8.854187817e-12; // vacuum permittivity in F/m 
constexpr double C0 = 2.99792458e8; // Speed of light in vacuum in m/s
namespace CSA{
    /*
    *******************************************************
    *            Necessary Conditons Chheck               *
    *******************************************************
    */
    // Function to check whether the curve is monotone decreasing or increasing    
    bool is_monotone(const Eigen::ArrayXd& g, MonotoneType curve_type){
        size_t m = g.size(); // Size of the curve

        if (m < 2){
            throw std::invalid_argument("Not enough spline knots for further processing");
        }

        if (curve_type == MonotoneType::Decreasing){
            return (g.tail(m-1) <= g.head(m-1)).all();
        }
        else if (curve_type == MonotoneType::Increasing){
            return (g.tail(m-1) >= g.head(m-1)).all();
        }
        else {
            throw std::invalid_argument("Invalid monotonicity type");
        }

    }

    // Function to return degree of monotonicity
    double degree_monotone(const Eigen::ArrayXd& g_left, const Eigen::ArrayXd& g_right){
        size_t m = g_left.size();
        Eigen::ArrayXd dx_left = g_left.bottomRows(m-1) - g_left.topRows(m-1);

        m = g_right.size();
        Eigen::ArrayXd dx_right = g_right.bottomRows(m-1) - g_right.topRows(m-1);

        double degree =  2*m - 0.5*((dx_left > 0).count() + (dx_right < 0).count()); // Shouldnt be zero

        return degree;
    }

    // Function to check whether the curve is convex
    bool is_convex(const Eigen::ArrayXd& g){
        size_t m = g.size();

        // Check if the vector has at least 3 elemen
        if (m < 3) {
            throw std::invalid_argument("Not enough spline knots for further processing");
        }

        // Calculate the second differences
        Eigen::ArrayXd dx2(m - 2);
        // segment(start idx, length)
        dx2 = g.segment(2, m-2) 
            - 2.0 * g.segment(1, m-2) 
            + g.segment(0, m-2);

        // Check if all second differences are non-negative
        return (dx2 >= 0).all();
    }

    /*
    *******************************************************
    *            Potential & Potential Coeffs             *
    *******************************************************
    */
    Eigen::ArrayXd calculate_potential_coeffs(const double& V0,
        const double& width_micrstr,
        const double& space_bw_strps, 
        const double& hw_arra, 
        const int& num_fs, 
        const Eigen::ArrayXd& g_left, 
        const Eigen::ArrayXd& x_left,
        const Eigen::ArrayXd& g_right, 
        const Eigen::ArrayXd& x_right){
        // Validate arguments
        //============================
        if (g_left.size() != x_left.size()){
            throw std::invalid_argument(std::format("Dimensions of x-axis vector and g-point vector for left side do not match! g: {}, x: {}", 
                g_left.size(), x_left.size()));
        }

        if (g_right.size() != x_right.size()){
            throw std::invalid_argument(std::format("Dimensions of x-axis vector and g-point vector for right side do not match! g: {}, x: {}", 
                g_right.size(), x_right.size()));
        }

        // Repeating or Constant terms
        //============================
        double M = g_left.size();
        double N = g_right.size();
        double d = space_bw_strps/2;

        Eigen::ArrayXd n = (Eigen::ArrayXd::LinSpaced(num_fs, 1, num_fs)); // nx1

        Eigen::ArrayXd alpha = n*PI/hw_arra; // nx1

        Eigen::ArrayXd m = (g_left.bottomRows(M-1) - g_left.topRows(M-1)) /
                        (x_left.bottomRows(M-1) - x_left.topRows(M-1)); // M-1x1
        Eigen::ArrayXd m_prime = (g_right.bottomRows(N-1) - g_right.topRows(N-1)) /
                        (x_right.bottomRows(N-1) - x_right.topRows(N-1)); // N-1x1

        double outer_coeff = 2*V0/hw_arra;

        // vn1
        //=========================
        // This calculation results in 2D so the .array() should be maped to ArrayXXd datatype , here a final .matrix will do the job
        Eigen::MatrixXd sin_left = ((alpha.matrix() * x_left.bottomRows(M-1).matrix().transpose()).array().sin()
                                    - (alpha.matrix() * x_left.topRows(M-1).matrix().transpose()).array().sin()).matrix(); // nx1 x 1xM-1 = nxM-1
        
        Eigen::ArrayXd vn1 = (1/alpha.square())*(
            (sin_left * m.matrix()).array()
        ); // nx1 x (nxM-1 x M-1x1) = nx1

        // vn2
        //=========================
        Eigen::ArrayXd vn2 = (1/alpha)*(
            ( 
            ((alpha.matrix() * x_left.topRows(M-1).matrix().transpose()).array().cos()).matrix()
            * g_left.topRows(M-1).matrix()
            ).array()
            - ( 
            ((alpha.matrix() * x_left.bottomRows(M-1).matrix().transpose()).array().cos()).matrix()
            * g_left.bottomRows(M-1).matrix()
            ).array()
        ); // nx1 x (cos(nx1 x 1XM-1) x M-1x1) = nx1

        // vn3
        //=========================
        Eigen::ArrayXd vn3 = (1/alpha)*(
            (alpha*d).cos() - (alpha*(d+width_micrstr)).cos()
        ); // nx1

        // vn4
        //=========================
        Eigen::MatrixXd sin_right = ((alpha.matrix() * x_right.bottomRows(N-1).matrix().transpose()).array().sin()
                                    - (alpha.matrix() * x_right.topRows(N-1).matrix().transpose()).array().sin()).matrix(); // nx1 x 1xM-1 = nxM-1
        Eigen::ArrayXd vn4 = (1/alpha.square())*(
            (sin_right.matrix() * m_prime.matrix()).array()
        ); // nx1 x (nxN-1 x N-1x1) = nx1

        // vn5
        //=========================
        Eigen::ArrayXd vn5 = (1/alpha)*(
            ( 
            ((alpha.matrix() * x_right.topRows(N-1).matrix().transpose()).array().cos()).matrix()
            * g_right.topRows(N-1).matrix()
            ).array()
            - ( 
            ((alpha.matrix() * x_right.bottomRows(N-1).matrix().transpose()).array().cos()).matrix()
            * g_right.bottomRows(N-1).matrix()
            ).array()
        ); // nx1 x (cos(nx1 x 1XN-1) x M-1x1) = nx1

        Eigen::ArrayXd vn = outer_coeff*(vn1+vn2+vn3+vn4+vn5);

        return vn; // nx1
    }
    
    Eigen::ArrayXd calculate_potential(const double& hw_arra,
        Eigen::ArrayXd& vn, 
        std::vector<double>& x){
        
        // Input potential coefficients must not be empty
        if(vn.rows() == 0){
            throw std::invalid_argument("Potenntial coefficients vn is empty");
        } 

        size_t num_fs = vn.size();
        Eigen::ArrayXd n = (Eigen::ArrayXd::LinSpaced(num_fs, 1, num_fs)); // nx1

        Eigen::ArrayXd alpha = n*PI/hw_arra; // nx1

        Eigen::MatrixXd x_vals = Eigen::Map<const Eigen::MatrixXd>(x.data(), x.size(), 1); // Mx1

        Eigen::ArrayXd sin = (x_vals * alpha.matrix().transpose()).array().sin(); // Mx1 x 1xn =  Mxn

        Eigen::ArrayXd VF = (sin.matrix() * vn.matrix()).array(); // Mxn x nx1 = Mx1

        return VF;
    }

    /*
    *******************************************************
    *                      Energy                         *
    *******************************************************
    */
   // sinh = (e^x - e^(-x))/2 => ln(sinh(x)) = ln(e^x - e^(-x)) - ln(2)
   // cosh = (e^x + e^(-x))/2 => ln(cosh(x)) = ln(e^x + e^(-x)) - ln(2)
   Eigen::ArrayXd logsinh(const Eigen::ArrayXd& vector){
        Eigen::ArrayXd absolute_vector = vector.abs();

        // Take care of overflow with threshold
        Eigen::ArrayXd logsinh_result = (absolute_vector > 33.0).select(
            absolute_vector - std::log(2.0), 
            (vector.exp() - (-vector).exp()).log() - std::log(2.0)); // nx1
            
        return logsinh_result;
   }

    Eigen::ArrayXd logcosh(const Eigen::ArrayXd& vector){
        Eigen::ArrayXd absolute_vector = vector.abs();

        // Take care of overflow with threshold
        Eigen::ArrayXd logcosh_result = (absolute_vector > 33.0).select(
            absolute_vector - std::log(2.0), 
            (vector.exp() + (-vector).exp()).log() - std::log(2.0)); // nx1
            
        return logcosh_result;
    }

    double calculate_energy(const double& er1,
        const double& er2,
        const double& hw_arra,
        const double& ht_arra,
        const double& ht_subs,
        const int& num_fs,
        Eigen::ArrayXd& vn){
        
        // Repeating or Constant terms
        //============================
        double e1 = er1 * E0;
        double e2 = er2 * E0;
        Eigen::ArrayXd n = (Eigen::ArrayXd::LinSpaced(num_fs, 1, num_fs)); // nx1
        
        Eigen::ArrayXd alpha = n*PI/hw_arra; // nx1
        
        Eigen::ArrayXd coeff = (n*PI/4)*vn.square(); // nx1

        // W1
        //=========
        Eigen::ArrayXd theta1 = alpha*(ht_arra-ht_subs); // nx1
        Eigen::ArrayXd coth1 = (logcosh(theta1) - logsinh(theta1)).exp();
        double w1 = e1 * coeff.matrix().dot(coth1.matrix()); // nx1 . nx1 = 1

        // W2
        //=========
        Eigen::ArrayXd theta2 = alpha*ht_subs; // nx1
        Eigen::ArrayXd coth2 = (logcosh(theta2) - logsinh(theta2)).exp();
        double w2 = e2 * coeff.matrix().dot(coth2.matrix()); // nx1 . nx1 = 1

        double w12 = w1 + w2;

        return w12;
    }

    /*
    *******************************************************
    *      Capacitance, Impedance, Epsilon Effective      *
    *******************************************************
    */
    double calculate_capacitance(double V0, double W){
        return (2*W)/(V0*V0);
    }

    double calculate_impedanceL(double capacitanceL){
        return 376.62*E0/capacitanceL;
    }

    double calculate_impedanceD(double capacitanceD, double capacitanceL){
        return 376.62*E0/(std::sqrt(capacitanceL*capacitanceD));
    }

    double calculate_epsilonEff(double capacitanceD, double capacitanceL){
        return capacitanceD/capacitanceL;
    }
} // namespace CSA

File: src/even/coupledstrip_lib.py
#################################################
# Author        : Philip Varghese Modayil
# Last Update   : 28-07-2025
# Topic         : Coupled Strip Lib Fucntions
# Description   : This file sets up the required functions for various calculation regarding the coupled microstrip
#################################################

#####################################################################################
#                                     Imports
#####################################################################################
from dataclasses import dataclass
from typing import Literal

from numba import njit
import numpy as np
from numpy.typing import NDArray

######################################################################################
#                            Couple Strip Arrangement
######################################################################################
@dataclass(frozen=False)
class CoupledStripArrangement:
    V0: float # Potential of the sytem, used to scale the system which is defaulted at V0=1.0
    hw_arra: float # half width of the arrangement, parameter a
    ht_arra: float # height of the arrangement, parameter b
    ht_subs: float # height of the substrate, parameter h
    space_bw_strps: float # gap between the two microstrips, parameter s
    width_micrstr: float # width of the microstrip, parameter w
    ht_micrstr: float # height of the microstripm, parameter t
    er1: float # dielectric constatnt for medium 1
    er2: float # dielctric constant for medium 2
    num_fs: int # number of fourier series coefficients
    num_pts: int # number of points for the piece wise linear approaximation
    mode: str # Even or Odd mode

######################################################################################
#                            Necessary Conditions Check
######################################################################################
@njit
def is_monotone(g: NDArray[np.float64], type: Literal["decreasing","increasing"]) -> bool:
    """
    Function to check for monotonicity, type  determines whether monotone decreasing or increasing

    Parameters
    ----------
    g : NDArray[np.float64]
        g points   for Piece Wise Linear function approximation
    type : Literal["decreasing","increasing"]
        flag to check for increasing or decreasing

    Returns
    -------
    bool
        returns True if monotone
    """
    dx: NDArray[np.float64] = np.diff(np.ascontiguousarray(g))
    if type == "decreasing":
        return bool(np.all(dx < 0))
    return bool(np.all(dx > 0)) 

@njit
def degree_monotonicity(g: NDArray[np.float64], type: Literal["decreasing","increasing"]) -> int:
    """
    Function to calculate the degree of monotonicity for the curve
    How many points follow the constraint

    Parameters
    ----------
    g : NDArray[np.float64]
        g points   for Piece Wise Linear function approximation
    type : Literal[&quot;decreasing&quot;,&quot;increasing&quot;]
        flag to check for increasing or decreasing

    Returns
    -------
    int
        number of points that follow the constraint
    """
    dx: NDArray[np.float64] = np.diff(np.ascontiguousarray(g))
    if type == "decreasing":
        return int(np.sum(dx < 0))
    return int(np.sum(dx > 0))

@njit   
def is_convex(g: NDArray[np.float64]) -> bool:
    """
    Function to check for convexity of the Piece Wise Linear function


    Parameters
    ----------
    g : NDArray[np.float64]
        g points   for Piece Wise Linear function approximation

    Returns
    -------
    bool
        returns True if convex

    Raises
    ------
    ValueError
        raise error  for insufficient number of g points
    """
    # Check if the array has at least 3 elements
    if len(g) < 3:
        raise ValueError("The array should have at least 3 elements to perform convexity check.\nPlease use more g-points")
    
    # Calculate the second differences
    g_contig: NDArray = np.ascontiguousarray(g)
    dx2 = g_contig[2:] - 2 * g_contig[1:-1] + g_contig[:-2]
    
    # Check if all second differences are non-negative
    return bool(np.all(dx2 >= 0))

@njit
def degree_convexity(g: NDArray[np.float64]) -> int:
    """
    Function to calculate the degree of convexity

    Parameters
    ----------
    g : g points   for Piece Wise Linear function approximation

    Returns
    -------
    int
        number of points that follow the constraint
    """
    # Calculate the second differences
    g_contig: NDArray = np.ascontiguousarray(g)
    dx2 = g_contig[2:] - 2 * g_contig[1:-1] + g_contig[:-2]
    
    # count number of positive values
    return int(np.sum(dx2>0))
######################################################################################
#                              Potential & Potential Coeffs
######################################################################################
@njit
def calculate_potential_coeffs(V0: float,
                               hw_arra: float,
                               width_micrstr: float,
                               space_bw_strps: float,
                               num_fs: int,
                               g_left: NDArray[np.float64], 
                               x_left: NDArray[np.float64],
                               g_right: NDArray[np.float64], 
                               x_right: NDArray[np.float64]) -> NDArray[np.float64]:
    """
    Function to calculate the Fourier potential coefficients for the Fourier series approximation

    Parameters
    ----------
    V0 : float
        potential at the microstrip, used to scale the equations
    hw_arra : float
        half width of the arrangement
    width_micrstr : float
        width of the microstrip
    space_bw_strps : float
        gap between the microstrip pair
    num_fs : int
        number of Fourier coefficients
    g_left : NDArray[np.float64]
        PWL g points for 0 <= x <= space_bw_strps/2
    x_left : NDArray[np.float64]
        x coordinates for g_left
    g_right : NDArray[np.float64]
        PWL g points for space_bw_strps/2 + width_micrstr <= x <= hw_arra
    x_right : NDArray[np.float64]
        x coordinates for g_right

    Returns
    -------
    NDArray[np.float64]
        Fourier coefficients

    Raises
    ------
    ValueError
        raises error if the dimensions of the left coordinates don't match
    ValueError
        raises error if the dimensions of the right coordinates don't match
    """
    # Dimensionality check
    if np.size(x_left) != np.size(g_left):
        raise ValueError(f'Dimensions of x-axis vector and g-points vector for left side do not match!\n\
                         size x_left:{np.size(x_left)}, size g_right:{np.size(g_left)}')  
    if np.size(x_right) != np.size(g_right):
        raise ValueError(f'Dimensions of x-axis vector and g-points vector for right side do not match!\n\
                         size x_right:{np.size(x_right)}, size g_right:{np.size(g_right)}')    
        
    # Repeating or constant terms
    #############################
    M: float = np.size(g_left)
    N: float = np.size(g_right)
    d: float = space_bw_strps/2
    
    n: NDArray[np.int64] = np.arange(0,num_fs) # 1xn
    
    beta: NDArray[np.float64] = ((2*n+1)*np.pi/(2*hw_arra)).astype(np.float64) # 1xn
    m: NDArray[np.float64] = (g_left[1:M] - g_left[0:M-1])/(x_left[1:M] - x_left[0:M-1]) # 1xM-1
    m_prime: NDArray[np.float64] = (g_right[1:N] - g_right[0:N-1])/(x_right[1:N] - x_right[0:N-1]) # 1xN-1
    
    x_left_vec: NDArray[np.float64] = np.reshape(np.ascontiguousarray(x_left),(-1,1)) # Mx1
    x_right_vec: NDArray[np.float64] = np.reshape(np.ascontiguousarray(x_right),(-1,1)) # Nx1
    
    outer_coeff: float = 2*V0/hw_arra
    
    # vn1
    ######
    vn1: NDArray[np.float64] = (1/beta**2)*(
        np.dot(np.ascontiguousarray(m), np.ascontiguousarray(np.cos(beta*x_left_vec[1:M]) - np.cos(beta*x_left_vec[0:M-1])))
    ) # 1xn x [1xM-1 x M-1xn] = 1xn
    
    # vn2
    ######
    vn2: NDArray[np.float64] = (1/beta)*(
        np.dot(np.ascontiguousarray(g_left[1:M]), np.ascontiguousarray(np.sin(beta*x_left_vec[1:M])))
        - np.dot(np.ascontiguousarray(g_left[0:M-1]), np.ascontiguousarray(np.sin(beta*x_left_vec[0:M-1])))
    ) # 1xn x [1xM-1 x M-1xn] = 1xn
    
    # vn3
    ######
    vn3: NDArray[np.float64] = (1/beta)*(np.sin(beta*(d+width_micrstr)) - np.sin(beta*d))
    
    # vn4
    ######
    vn4: NDArray[np.float64] = (1/beta**2)*(
        np.dot(np.ascontiguousarray(m_prime), np.ascontiguousarray(np.cos(beta*x_right_vec[1:N]) - np.cos(beta*x_right_vec[0:N-1])))
    ) # 1xn x [1xN-1 x N-1xn] = 1xn
    
    # vn5
    ######
    vn5: NDArray[np.float64] = (1/beta)*(
        np.dot(np.ascontiguousarray(g_right[1:N]), np.ascontiguousarray(np.sin(beta*x_right_vec[1:N])))
        - np.dot(np.ascontiguousarray(g_right[0:N-1]), np.ascontiguousarray(np.sin(beta*x_right_vec[0:N-1])))
    ) # 1xn x [1xN-1 x N-1xn] = 1xn
    
    # vn
    ######
    vn: NDArray[np.float64] = outer_coeff*(vn1+vn2+vn3+vn4+vn5)
    
    return vn.astype(np.float64)

@njit
def calculate_potential(hw_arra: float,
                        vn: NDArray[np.float64],
                        x: NDArray[np.float64]) -> NDArray[np.float64]:
    """
    Function to calculate the potential with Fourier reconstruction

    Parameters
    ----------
    hw_arra : float
        half width of the arrangement
    vn : NDArray[np.float64]
        Fourier coefficients
    x : NDArray[np.float64]
        x axis coordinates

    Returns
    -------
    NDArray[np.float64]
        Potential Fourier reconstruction for the given x
    """
    num_fs: int = np.size(vn)
    
    n: NDArray[np.int64] = np.ascontiguousarray(np.arange(0,num_fs))[:, np.newaxis] # nx1
    beta: NDArray[np.float64] = ((2*n+1)*np.pi/(2*hw_arra)).astype(np.float64) # nx1
    cos: NDArray[np.float64] = np.cos(beta*x) # nxm
    VF: NDArray[np.float64] = np.dot(np.ascontiguousarray(vn),np.ascontiguousarray(cos)) # 1xn x nxm = 1xm
    
    return VF.astype(np.float64)
######################################################################################
#                                        ENERGY
######################################################################################
# define logarithmic implementations for cosh and sinh to improve numerical stability
# sinh = (e^x - e^(-x))/2 => ln(sinh(x)) = ln(e^x - e^(-x)) - ln(2)
# cosh = (e^x + e^(-x))/2 => ln(cosh(x)) = ln(e^x + e^(-x)) - ln(2)
@njit
def logsinh(vector: NDArray[np.float64]) -> NDArray[np.float64]:
    """
    Function to calculate the logsinh with numerical stability via thresholding

    Parameters
    ----------
    vector : NDArray[np.float64]
        theta in sinh(theta)

    Returns
    -------
    NDArray[np.float64]
        log(sinh)
    """
    absolute_vector: NDArray[np.float64] = np.abs(vector)

    logsinh_result: NDArray[np.float64] = np.where(absolute_vector > 33.0,
                                                   absolute_vector - np.log(2),
                                                   np.log(np.exp(vector) - np.exp(-vector)) - np.log(2))
    return logsinh_result

@njit
def logcosh(vector: NDArray[np.float64]) -> NDArray[np.float64]:
    """
    Function to calculate the logcosh with numerical stability via thresholding

    Parameters
    ----------
    vector : NDArray[np.float64]
        theta in cosh(theta)

    Returns
    -------
    NDArray[np.float64]
        log(cosh)
    """
    absolute_vector: NDArray[np.float64] = np.abs(vector)

    logcosh_result: NDArray[np.float64] = np.where(absolute_vector > 33.0,
                                                   absolute_vector - np.log(2),
                                                   np.log(np.exp(vector) + np.exp(-vector)) - np.log(2))
    return logcosh_result

@njit
def calculate_energy(er1: float, er2: float, hw_arra: float, ht_arra: float, ht_subs: float, vn: np.ndarray) -> float:
    """
    Function to calculate the energy of coupled strip system odd mode

    Parameters
    ----------
    er1 : float
        relative permitivity of medium 1
    er2 : float
        relative permitivity of medium 2
    hw_arra : float
        half width of the arrangement
    ht_arra : float
        height of the arrangement
    ht_subs : float
        height of the substrate
    vn : np.ndarray
        Fourier coefficients

    Returns
    -------
    float
        energy of the system
    """
    # Constant terms
    ################
    e0: float = 8.854187817E-12
    e1: float = er1*e0 # 8.54E-12 is permittivity constant e0
    e2: float = er2*e0
    
    num_fs: int = np.size(vn)
    n: NDArray[np.int64] = np.arange(1,num_fs+1)
    
    # Energy Formula Odd-Mode
    #########################
    coeff = (n*np.pi/4)*vn**2 # 1 x n
    
    # w1
    #####
    theta1: NDArray[np.float64] = (n*np.pi*(ht_arra-ht_subs)/hw_arra).astype(np.float64) # 1 x n
    coth1: NDArray[np.float64] = np.exp(logcosh(theta1)-logsinh(theta1)) # log(cosh/sinh) = log(cosh) - log(sinh)
    w1: NDArray[np.float64] = e1*coth1 # 1xn

    # w2
    #####
    theta2: NDArray[np.float64] = (n*np.pi*ht_subs/hw_arra).astype(np.float64) # 1 x n
    coth2: NDArray[np.float64] = np.exp(logcosh(theta2)-logsinh(theta2)) # log(cosh/sinh) = log(cosh) - log(sinh)
    w2: NDArray[np.float64] = e2*coth2 # 1xn


    W: float = np.sum(coeff*(w1+w2))
    
    return W
    
######################################################################################
#                        Capacitance, Impedance, Epsilon Effective
######################################################################################
@njit
def calculate_capacitance(V0: float, W:float) -> float:
    """
    Fucntion to calculate capacitance of the system

    Parameters
    ----------
    V0 : float
        potential at the microstrip
    W : float
        energy of the system

    Returns
    -------
    float
        capacitance of the system
    """
    
    C: float = (2*W)/(V0**2)
    
    return C

@njit
def calculate_impedance(cD: float = 1.0, cL: float = 1.0, env: Literal["caseD", "caseL"] = "caseD") -> float:
    """
    Function to calculate the impedance of the system based on the environment case

    Parameters
    ----------
    cD : float
        capacitance of the caseD system, by default 1.0 (arbitrary no significance)
    cL : float
        capacitance of the caseL system, by default 1.0 (arbitrary no significance)
    env : Literal["caseD", "caseL"]
        case of the environment (caseD/caseL), by default "caseD"

    Returns
    -------
    float
        _description_
    """
    e0: float = 8.854187817E-12
    match env:
        case "caseD":
            return 376.62*e0/((cD*cL)**0.5)
        case "caseL":
            return 376.62*e0/cL
        case _:
            return 0.0 # must not happen as there is a default state(just for the type checker to pass)

@njit
def calculate_epsilonEff(cD: float, cL: float) -> float:
    """
    Function to calculate effective dielectric constant

    Parameters
    ----------
    cD : float
        case D capacitance
    cL : float
        case L capacitance

    Returns
    -------
    eps_eff: float
        effective dielectric constant
    """
    return cD/cL
######################################################################################
#                           Surface Charge Density
######################################################################################